### DISCUSSION TITLE
Big int doesn't work well with JavaScript

### DISCUSSION THREAD

**Discussion by dongyuzheng** [encode/django-rest-framework#8689] [Potential Issue]
Because JS numbers suffer from FPE.

Take this big integer ID for example:

![image](https://user-images.githubusercontent.com/6127032/194780151-a11df304-6508-47a3-a23c-b9d8b04fada8.png)

The returned ID is correct in the response json, but JS cannot load it properly.

This leads to nasty bugs.

The solution is to make the id into a `serializers.CharField()`: https://stackoverflow.com/questions/72943702/using-bigautofield-pk-from-drf-in-javascript ... but this took longer than it should've to debug and could've been more disasterous.

Thoughts on "how to auto handle or warn the user"? E.g., "if field is bigint and number might suffer from FPE: error until user sets REST_FRAMEWORK_NO_ERROR_BIGINT_JS_FPE"

Additional tags: int64 bigint pk id fpe "wrong value"

**Comment by lovelydinosaur** [comment]
Related...

* https://github.com/encode/django-rest-framework/issues/5007
* https://github.com/encode/django-rest-framework/issues/5006

Some numbers for reference...

* 2,147,483,647 - [Django Integer field safe database range](https://docs.djangoproject.com/en/4.1/ref/models/fields/#django.db.models.IntegerField)
* 9,007,199,254,740,991 - [JSON Interoperability range](https://www.rfc-editor.org/rfc/rfc7159#section-6)
* 9,223,372,036,854,775,807 - [Django BitInt field safe database range](https://docs.djangoproject.com/en/4.1/ref/models/fields/#bigintegerfield)

Anyways. Yes - the answer in that case is correct - if there's no need for numeric calculations and you're using that field as a PK, then a `CharField` for it makes sense. It's just an opaque identifier.

It's not obvious exactly if/what we should do about this in REST framework itself. Particularly without breaking backwards compat.
The error-unless-configured-to-allow you mention does make sense, but there's a bunch of details around that I'm unsure about.

**Reply by dongyuzheng** [reply]
Makes sense that handling this isn't really drf's responsibility. 

In lieu of code change, maybe we can add a warning about this somewhere very noticible in the docs, or even at multiple plqces? E.g., here https://www.django-rest-framework.org/tutorial/quickstart/#serializers

Warning: BigInt has known compatibility issues with JavaScript due to floating point inaccuracies in JS. A recommended workaround is to define `id = serializers.CharField()` in your Serializer. You can also enable this as default behaviour by setting `REST_FRAMEWORK_SERIALIZE_BIGINT_PK_AS_CHARFIELD = True`

**Reply by lovelydinosaur** [reply]
I'd probably be okay with a `REST_FRAMEWORK_BIGINT_AS_CHARFIELD` setting, sure.

(I don't think we'd limit it to the PK case.)
